File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.cpp
class Solution {
public:
    bool dp[5001][101];
    bool canPartition(vector<int>& nums) {
        dp[0][0] = true;
        int total = 0;
        for (int i=0; i<nums.size(); i++) {
            total += nums[i];
        }

        if (total % 2 ) {
            return false;
        }

        int target = total / 2;

        for (int s = 1; s <= target; s++) {
            for (int i=1; i<=nums.size(); i++) {
                dp[s][i] = dp[s][i-1];

                if (s - nums[i-1] >= 0) {
                    dp[s][i] |= dp[s - nums[i-1]][i-1];
                }
            }
        }

        return dp[target][nums.size()];
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        const prefix = Array(nums.length);
        const suffix = Array(nums.length);

        for (let i = 0; i<nums.length; i++) {
            if (i > 0) {
                prefix[i] = nums[i] * prefix[i-1];
            } else {
                prefix[i] = nums[i];
            }
        }

        for (let i = nums.length - 1; i >= 0; i--) {
            if (i < nums.length - 1) {
                suffix[i] = nums[i] * suffix[i+1];
            } else {
                suffix[i] = nums[i];
            }
        }

        const ans = [];

        for (let i=0; i<nums.length; i++) {
            const pre = i > 0 ? prefix[i-1] : 1;
            const suf = i < nums.length - 1 ? suffix[i+1] : 1;

            ans.push(pre*suf);
        }

        return ans;

    }
}
